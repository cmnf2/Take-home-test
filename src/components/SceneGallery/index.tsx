import React from 'react';

import SceneGallery from '@components/SceneGallery/sceneGallery';

export default (): JSX.Element => <SceneGallery />;
