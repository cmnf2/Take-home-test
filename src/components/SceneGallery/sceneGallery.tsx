import React, { ReactElement, useEffect } from "react";
import ImageList from "@mui/material/ImageList";
import ImageListItem from "@mui/material/ImageListItem";
import Modal from "@mui/material/Modal";

import { fetchScenes } from "../helpers/fetchScenes";
import {randomizeSvgColors} from "../helpers/randomizeSVGColors";

import style from "./styles.scss";
import "../../assets/images/loader.gif";

export interface Scene {
  key: string;
  url: string;
  propCat: string;
  keywordKeys: string[];
}

function SceneGallery(): ReactElement {
  const loader: Scene[] = [
    {
      key: "loader",
      url: "/assets/images/loader.gif",
      propCat: "loadinggif",
      keywordKeys: ["keyord1"],
    },
  ];

  const [scenes, setScenes] = React.useState(loader);
  const [bigImage, setBigImage] = React.useState("");
  const [open, setOpen] = React.useState(false);

  const closeModal = () => {
    setOpen(false);
  };

  useEffect(() => {
    if (scenes.length > 0 && scenes == loader) {
      loadScenes();
    }
  });

  const spotlightImage = async (url: string) => {
    try {
      const image = await fetch(url).then((res) => res.text());
      setBigImage(randomizeSvgColors(image));

      setOpen(true);
    } catch (e) {
      // An error ocurred, don't open the modal
      // log the error
      console.log(e);
    }
  };

  const loadScenes = async () => {
    try {
      const scenes = await fetchScenes();

      setScenes(scenes.data.sceneGroups as Scene[]);
    } catch (e) {
      // Empty loaders and display error message
      setScenes([]);
    }
  };

  return (
    <div className="sceneGallery">
      <h1 className={style.SceneGalleryHeader}>Available Scenes</h1>

      <ImageList rowHeight={"auto"} className={style.SceneList}>
        {scenes.length > 0 ? (
          scenes.map((scene) => (
            <ImageListItem
              key={scene.key}
              className={style.Scene}
              role="button"
              onClick={() =>
                scene.key == "loader" ? {} : spotlightImage(scene.url)
              }
            >
              <img src={scene.url} alt={scene.propCat} loading="lazy" />
            </ImageListItem>
          ))
        ) : (
          <p id="ErrorMessage" role="alert">
            We're sorry, something went wrong. Please try back another time.
          </p>
        )}
      </ImageList>

      <Modal
        open={open}
        onClose={closeModal}
        aria-describedby="modal-modal-description"
      >
        <div className={style.Spotlight}>
          <button onClick={closeModal}>
            close
          </button>
          <div
            className={style.SpotlightImage}
            dangerouslySetInnerHTML={{ __html: bigImage }}
          ></div>
        </div>
      </Modal>
    </div>
  );
}

export default SceneGallery;
