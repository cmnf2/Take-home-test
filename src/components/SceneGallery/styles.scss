.SceneGalleryHeader {
  text-align: center;
}

// SCENES
ul.SceneList {
  max-width: 1020px;
  width: 100%;
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;

  @media screen and (min-width: 800px) {
    margin: 0 auto;
  }
}

.Scene {
  width: 100%;
  box-sizing: border-box;
  cursor: pointer;

  @media screen and (min-width: 400px) {
    width: 48%;
    margin: 0.5%;
  }

  @media screen and (min-width: 800px) {
    width: 30%;
    margin: 0.5%;
  }

  @media screen and (min-width: 1200px) {
    width: 21%;
    margin: 0.5%;
  }
}

// MODAL
.Spotlight {
  display: flex;
  justify-content: center;
  align-content: baseline;
  margin-top: 3rem;
  position: relative;
  box-sizing: border-box;
  &:focus {
    outline: none;
  }

  button {
    position: absolute;
    top: 0.5rem;
    border: none;
    cursor: pointer;
    font-size: 1.2rem;
    background-color: white;
  }
}

.SpotlightImage {
  background-color: white;
  min-height: 5rem;

  svg {
    padding: 0.5rem;
  }

  p {
    width: 80%;
  }
}
