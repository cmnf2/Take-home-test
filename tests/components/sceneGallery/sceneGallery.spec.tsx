import React from 'react';
import renderer from 'react-test-renderer';
import fetchMock from "jest-fetch-mock";

import SceneGallery from '@components/SceneGallery';

const testResponse = [
  {
    key: "testImage",
    url: "/assets/images/test.gif",
    propCat: "random",
    keywordKeys: ["keyword1"],
  }
];

fetchMock.enableMocks();
let scenesGallery: any;

describe('SceneGallery', () => {
  beforeAll(async () => {
    fetchMock.doMock()
    fetchMock.mockResponseOnce(JSON.stringify(testResponse));

    scenesGallery = await renderer.create(<SceneGallery />);
  });

  test('should render component properly', () => {
    fetchMock.mockResponseOnce(JSON.stringify(testResponse));

    const tree = scenesGallery.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
