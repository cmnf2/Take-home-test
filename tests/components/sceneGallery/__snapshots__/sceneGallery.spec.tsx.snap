// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneGallery should render component properly 1`] = `
<div
  className="sceneGallery"
>
  <h1
    className="SceneGalleryHeader"
  >
    Available Scenes
  </h1>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-rvv586-MuiImageList-root{display:grid;overflow-y:auto;list-style:none;padding:0;-webkit-overflow-scrolling:touch;}",
      }
    }
    data-emotion="css rvv586-MuiImageList-root"
  />
  <ul
    className="MuiImageList-root MuiImageList-standard SceneList css-rvv586-MuiImageList-root"
    style={
      Object {
        "gap": 4,
        "gridTemplateColumns": "repeat(2, 1fr)",
      }
    }
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-pxxd1-MuiImageListItem-root{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;position:relative;line-height:0;-webkit-flex-direction:column;-ms-flex-direction:column;flex-direction:column;}.css-pxxd1-MuiImageListItem-root .MuiImageListItem-img{object-fit:cover;width:100%;height:auto;-webkit-box-flex:1;-webkit-flex-grow:1;-ms-flex-positive:1;flex-grow:1;}",
        }
      }
      data-emotion="css pxxd1-MuiImageListItem-root"
    />
    <li
      className="MuiImageListItem-root MuiImageListItem-standard Scene css-pxxd1-MuiImageListItem-root"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "gridColumnEnd": "span 1",
          "gridRowEnd": "span 1",
          "height": "auto",
          "marginBottom": undefined,
        }
      }
    >
      <img
        alt="random"
        className="MuiImageListItem-img"
        loading="lazy"
        src="/assets/images/loader.gif"
      />
    </li>
  </ul>
</div>
`;
